trustee: Configure reference values based on all possible PCR combinations

bgartzi opened the pull request:
Now the compute-pcrs library supports computing PCR value combinations. In other words, given all events from image A and image B (where some of the components got updated), it computes all possible PCRs of the intermediate states that a node could go through during an update from image A to image B.

While implementing this I found a compute-pcrs bug that this PR relies on: https://github.com/trusted-execution-clusters/compute-pcrs/pull/56

For now, I added a workaround commit that is not signed-off-by me. Once the compute-pcrs library gets merged, I will update this PR.

Another discussion topic that comes to my mind is how unit tests are implemented to face this new integration. They sure could be way more isolated from compute-pcrs, which I will update if relevant.

Jakob-Naucke reviewed (commented):
Very nice. A few comments.

Jakob-Naucke commented on attestation-key-register/Containerfile:
```diff
@@ -3,7 +3,7 @@
 
 ARG build_type
-FROM ghcr.io/trusted-execution-clusters/buildroot AS builder
+FROM ghcr.io/trusted-execution-clusters/buildroot:fedora AS builder
```
`:latest` and `:fedora` point to Fedora 43 images that are apparently different, but I don't even know how they're different. Is there a difference that matters to this PR?

Jakob-Naucke commented:
/retest

Jakob-Naucke commented on operator/src/trustee.rs:
```diff
@@ -72,22 +74,28 @@ pub fn get_image_pcrs(image_pcrs_map: ConfigMap) -> Result<ImagePcrs> {
             .entry(format!("pcr{}", pcr.id))
             .or_default()
-            .push(JsonString(pcr.value.clone()));
+            .insert(hex::encode(pcr.value.clone()));
```
nit: no need to clone `hex::encode` args, it takes `AsRef` (also two in `get_image_pcrs_success`)

Jakob-Naucke reviewed (commented):
> Without proper warmup, `test_combined_image_pcrs_configmap_updates` is still a bit flaky due to the 180s timeout in `verify_expected_pcrs`. I could just increase it to 300 and call it good, but I might end up adding the secondary approved image to the images that are pre-pulled to the cluster, which might be a better thing to do.

I agree this is nicer because it decouples timeouts from the test host's internet speed. I noticed we don't even pre-pull in CI at all yet. Could you add it to `.github/workflows/integration-tests.yml` before the integration tests?

Jakob-Naucke commented on scripts/pre-pull-images.sh:
```diff
@@ -8,11 +8,25 @@
+index=0
+pull_approved_image "$APPROVED_IMAGE" $index
+
+for image in $EXTRA_APPROVED_IMAGES; do
```
I'm not a fan of the extra env. I'd rather keep these out of the Makefile, put a const for the test image at the top of `tests/trusted_execution_cluster.rs`, and grep-sed for that here, WDYT?


```suggestion
combined_pcrs_test_image=$(grep "const COMBINED_PCRS_TEST_IMAGE" tests/trusted_execution_cluster.rs | \
                           sed -E 's/.*"(.*)";/\1/')
for image in "$APPROVED_IMAGE" "$combined_pcrs_test_image"; do
```

bgartzi replied:
skipped the for, but I can put it there. You tell me.

Jakob-Naucke reviewed (commented):
urgh, two start failures in the test (probably not your fault) and a stop failure (fix is coming soon). I'll try and improve these soon.

Jakob-Naucke reviewed (approved):
no, I like it. @alicefr PTAL too bc it's pretty big

alicefr commented on api/trusted-cluster-gen.go:
```diff
@@ -184,34 +195,33 @@ func generateTrustedExecutionClusterCR(args *Args) error {
+				Kind:       "ApprovedImage",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("coreos-%d", i),
```
can we use something different then the index? maybe the instead the approvedImage with the `/` and `:`replaced by `-`. Or do we have a limit in the name lenght?

alicefr replied:
Also `coreos` wasn't great by the way

bgartzi replied:
Wrote a some tweaks on trusted-cluster-gen and things that use it to allow configuring approved image references together with their names.
Then I tried giving approved images more specific names.

alicefr commented on test_utils/src/constants.rs:
```diff
@@ -0,0 +1,233 @@
+macro_rules! expected_pcr7_hash {
+    () => {{ "b3a56a06c03a65277d0a787fcabc1e293eaa5d6dd79398f2dda741f7b874c65d" }};
+}
+
```
One question, do we expect these constant to remain always the same or do we need to modify them for example when we upgrade the fedora version of the approved image?

bgartzi replied:
They might change with approved images, specially pcr4. If we'd like to parametrize this so we could hardcode this upper in the stack, such as in the makefile, the problem wouldn't be passing pcr values, but the events of each of them.

bgartzi replied:
Got rid of unnecessary macros and put those values into constants with more clear names. Then, yes, I keep using some macros as helpers to build tpmevents out of constant values that should be now easier to update.

alicefr commented on tests/trusted_execution_cluster.rs:
```diff
@@ -353,3 +353,69 @@ async fn test_approved_image_readoption() -> anyhow::Result<()> {
+named_test! {
+async fn test_combined_image_pcrs_configmap_updates() -> anyhow::Result<()> {
+    let test_ctx = setup!([
+        "quay.io/trusted-execution-clusters/fedora-coreos@sha256:372a5db90a8695fafc2869d438bacd7f0ef7fd84f63746a450bfcd4b8b64ae83",
```
Question, how do we upgrade this? Should it make sense to move the definition out of the test and put an env variable. Same question, how do we get the expected pcr value.

Jakob-Naucke replied:
I mean, it's just a second version for testing that combination works, we can probably leave it for a few years. It's not a moving target like FCOS versions.

bgartzi replied:
As soon as we want to write another test that handles 2 approved images and their upgrade path, yes. We can get ahead of time and put it that way if you want. I don't have a personal choice.

When it comes to upgrades, if we would like to change the image to a completely new one, we need to boot a VM with that image, check the tpm event log and look for shim, grub and kernel related events. Then we need to compute possible combinations, which include primary bootloader + secondary kernel and secondary bootloader + primary kernel.

alicefr replied:
can we document in a comment how we build this image and describe how we extracted the pcrs values. At least, we won't forget it

bgartzi replied:
Added a document in `docs/dev/test_approved_images.md` trying to explain what each image carry, and some insights on how to calculate the pcrs, and combined pcrs upon the event of an approved image update.

alicefr commented on tests/trusted_execution_cluster.rs:
```diff
@@ -22,6 +22,8 @@ use trusted_cluster_operator_lib::{
 use trusted_cluster_operator_test_utils::constants::*;
 use trusted_cluster_operator_test_utils::*;
 
+const COMBINE_PCRS_UPDATE_TEST_IMAGE: &str = "quay.io/trusted-execution-clusters/fedora-coreos@sha256:372a5db90a8695fafc2869d438bacd7f0ef7fd84f63746a450bfcd4b8b64ae83";
```
This is make sense to squash the definition of this image in the commit where this was introduced

bgartzi replied:
Did you also mean the pre-pull logic? Just the const definition, right?

alicefr replied:
Just the const definition

bgartzi committed:
workspace: Add hex as a dependency

And update those crates that were making use of it. Later we will also
use it in some other crates.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
trustee.tests: Make dummy_pcrs a bit less dummy

trustee tests were relying on some values that didn't hold any TPMEvent
or any PCR value that related to them.

This commit turns those dummy values into something that are closer to
something we could expect in reality.

The main reason to do this is that the compute-pcrs logic does not like
empty event PCRs, as it's based on that to reconstruct them.

This goes against the purpose of a unit test, as we should actually mock
that external part so as not to rely on it, but decided this was simpler
at least as an initial proposal.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
trustee: Combine approved image PCR values

Update trustee reference values with all possible combinations between
approved PCR values, not just those that come directly from image PCRs.

This computes reference values for possible stages during node updates
in which a node could be booting some components from image A and some
other from image B.

This commit also adds a test to check that pcr4 is well covered in front
of bootloader and kernel updates. The test is closer to an integration
test than to an unit-test as it relies on values that are close to real
values, and how the compute-pcrs lib's combine_images processes them.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
bump: bring latest compute-pcrs in

I broke some of the compute-pcrs' lib APIs, so apart from just pointing
into a newer commit, we also need to tweak a couple of things here and
there.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
test_utils: Remove redundancy in verify_expected_pcrs

The logic checking that the image-pcrs configmap was populated moved
into using await_condition instead of a poller. Some conditions were
checked twice inside of the actual condition and subsequent assertions.
This commit removes those assertions in favor of the await call.

Moreover, it makes more sense from a stability point of view to wait for
the whole condition to happen, as it could be the case that a Nth image
is approved, and we run all the assertions (and fail) because only one
of the N images was found in the configmap. This way, the tests will
wait until the expected images of all approved images are present.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
trusted-cluster-gen: Support multiple images

Support the possibility of generating CRs for multiple approved images.
These are stored sepparately in different yaml files.

This changes the way trusted-cluster-gen names approved_image_cr.yaml
files, as it's not just one, but many of them that can be marshalled.
That's why test dependencies are updated to also be able to handle
multiple approved image CRs.

As we now deal with multiple images, and "coreos" is a bit to vague in
that case, approved-images now accept both the image reference and a
name, so we can identify them easily. Instead of calling "coreos" to the
existing one, we now call it "coreos-approved-primary".

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
tests, trusted_execution_cluster: Refactor expected base PCR values

These could be used by other tests too. Make it a macro so it can be
reused, and define a few more macros for standard TPMEvents and PCR
values that are constant across integration tests at the moment, such as
PCR7 and PCR14.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
tests, integration: PCR combination on bootloader+kernel update

Add an integration test in which 2 approved images with different
bootloader and kernel are added to the cluster. This emulates the
situation in which a coreos image could be undergoing a bootloader and
kernel update.

The test checks that 2 images are added to the image pcr config map, and
then checks that the reference values contain all possible pcr4
combinations. pcr7 and pcr14 are constant in this case, so there are not
combinations possible (apart from the original value).

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
.github: Pre-pull images before running integration tests

This help not hitting some test timeouts due to images not being pulled
quickly enough.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
scripts: Pre-pull extra approved images

Some integration tests might run on more approved images. Find which it
is based on the test file then pre-pull it.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>
Assisted-By: Claude Opus 4.6 <noreply@anthropic.com>

bgartzi committed:
test_utils: Verify expected pcrs based on TestContext

Refactor the logic that was verifying expected image pcrs in
test_image_pcrs_configmap_updates as a method of TestContext.

It also slightly changes the logic of the poller to stop when the length
of pcrs equals the one expected.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
compute-pcrs: Don't compute PCR7

The compute pcrs job was computing pcr7 based on the guest OS
information. However, PCR7 is mainly affected by the EFI variables,
which are configured by the platform, or virtualization host.

As we lack a proper way to retrieve the EFI vars of the platform that
nodes are running on, we don't attest pcr7, so we should just not
compute it.

This commit removes the bits that computed pcr7 in the compute-pcrs job
and those that checked it during integration tests. It leaves the test
logic covering pcr7 in operator/src/trustee.rs as it should be agnostic
to whatever pcr prediction comes in.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

bgartzi committed:
docs: Updating approved images and expected PCR values

Covering the primary and secondary approved images and also possible
combinations between the two of them.

Signed-off-by: Beñat Gartzia Arruabarrena <bgartzia@redhat.com>

Jakob-Naucke commented:
@dependabot unignore compute-pcrs-lib
